### DISCUSSION TITLE
Select a substring of a memo field without having the base field on the entity

### DISCUSSION THREAD

**Discussion by primozcerar** [linq2db/linq2db#3465] [Q&A] [answered]
I can't figure out how to map an entity property to a substring of a memo field, without having that field mapped to another property or selecting the whole memo field.
This is a performance issue as a substring of a memo field with for example first 200 characters is much faster to read in firebird than the whole memo field, even if its shorter.
I have tried ExpressionMethod without success as anything I try requires me to have the property for the field on the entity, Sql.Substring seems to only accept a string value, not a field name.
I know I can do this by selecting from the entity and creating a new instance but I would really like to have it mapped on the entity itself.
Is there a way to do this?

**Comment by MaceWindu** [comment]
This should work:

for explicit use
```cs
[Expression("SUBSTR(my_memo)", ServerSideOnly=true)]
public string Substring {get;set;}
```
for use as column
```cs
[ExpressionMethod(nameof(MemoExpr), IsColumn = true)]
public string Memo { get; set; }

static Expression<Func<Entity, string>> MemoExpr()
{
    e => Sql.Expr<string>($"substr(memo)");
}
```

**Reply by primozcerar** [reply]
@MaceWindu  This works. Thank you very much. There is one issue left. Is there a way to get the table alias to insert it into the substring expression?

**Reply by MaceWindu** [reply]
Sorry about it, forgot about alias. You can change it to this:
```cs
// Sql.TableField will handle aliasing
// Substring is already mapped to SQL by linq2db
return e => Sql.TableField<Entity, string>(e, "MemoColumn").Substring(0, 200);
```

**Reply by primozcerar** [reply]
@MaceWindu Sql.TableField seems to be internal and I can not call it.

**Reply by MaceWindu** [reply]
omg... wait a minute

**Reply by MaceWindu** [reply]
Let's see. we will fix it for next version (expose `Sql.TableField` or fix `Sql.Property` which currently fails). 

For now I propose to copy `Sql.TableField` implementation to your project as workaround:

```cs
private class TableFieldBuilder : IExtensionCallBuilder
		{
			public void Build(ISqExtensionBuilder builder)
			{
				var tableExpr = builder.GetExpression(0);
				var fieldName = builder.GetValue<string>(1);
				var sqlField = tableExpr as SqlField;

				if (sqlField == null)
					throw new LinqToDBException("Can not find Table associated with expression");

				builder.ResultExpression = new SqlField(sqlField.Table!, fieldName);
			}
		}

		[Extension("", BuilderType = typeof(TableFieldBuilder))]
		internal static TColumn TableField<TEntity, TColumn>([NoEnumeration] TEntity entity, string fieldName)
		{
			throw new LinqToDBException("'Sql.TableField' is server side only method and used only for generating custom SQL parts");
		}
```

**Reply by MaceWindu** [reply]
#3468

**Comment by primozcerar** [comment]
Copying the TableField implementation worked. Thank you very much. This will be useful for many situations I believe if it's made public.

**Comment by Usaga** [comment]
Actually, if you are using projections (which you definitely should be doing all the time), you can just use `Sql.Property` in your query, without having to copy any `Sql.TableField` code.

```C#
var result = context.GetTable<Table>()
   .Select(x => new ProjectionClass
   {
     ... other properties
     Memo = Sql.Property<string>(x, "SomeFieldName").Substring(0, 200)
   })
  .ToList();
```

This allows you to access any table field not added into entity. Even more, you can access fields of joined tables, which is more flexible than `Sql.TableField`.

Or, if you prefer fetch entire entities:
```C#
        [ExpressionMethod(nameof(SomeExpression), IsColumn = true)]
        [Column(SkipOnEntityFetch = true)]
        public string Memo{ get; set; }

        private static Expression<Func<CaseLog, string>> MemoExpression()
        {
            return x => Sql.Property<string>(x, "SomeFieldName").Substring(0, 2);
        }
```

**Comment by primozcerar** [comment]
@Usaga Thank you for that but unfortunately I'm still using Linq2db v2 and switching my project to v3 isn't trivial, so I don't have the SkipOnEntityFetch option. Also MaceWindu mentioned that Sql.Property does not work in an expression method but this might be due to the missing SkipOnEntityFetch.

**Comment by primozcerar** [comment, marked as answer]
@Usaga @MaceWindu I switched my project to 3.6.0 and Sql.Property seems to be working even without the [Column(SkipOnEntityFetch = true)] attribute. So it seems that Sql.TableField is not required.

Thank you both.
